### PR TITLE
Fgmres-dev// adding Inner IR for CUSOLVER_LU

### LINKED ISSUE #430: Add iterative refinement for LU linear solver from cuSOLVER library
Implement iterative refinement on GPU for the LU-based direct solver from cuSOLVER library. GPU direct solvers do not always produce high quality solution and the iterative refinement may be an efficient way to address that.

@cnpetra @nychiang

### PR DESCRIPTION
**Pull request by kswirydo**
What changed:
- FGMRES was added as iterative refinement to CUSOLVER_LU (fgmres has its own class, same cpp and hpp files as CUSOLVER_LU code)
- Options were added in hiopOptions for FGMRES IR (on/off, maximum iterations, restart frequency, tolerance, orthogonalization scheme).
- This method was tested with Sparse/NlpSparseEx1. It works with all orthogonalization methods.
Purpose:
- Add this code to develop to give users more flexibility choosing solver strategy. 

Fixes #430

### CODE REVIEW

**Review by pelesh — APPROVED** [review]

**Comment by pelesh** [comment]
@cnpetra, I rebased wrt current `develop`. I think it is safe to merge this PR.

File: src/LinAlg/hiop_cusolver_ir_custom_kernels.cu [hunk 1]
```diff
@@ -0,0 +1,208 @@
+//written by KS some time in 2019
```
Comment (cnpetra) [inline src/LinAlg/hiop_cusolver_ir_custom_kernels.cu:1]: needs HiOp copyright text

File: src/LinAlg/hiop_cusolver_ir_custom_kernels.cu [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  }
+  if(t == 0) {
+    result[blockIdx.x] = s_tmp1[0];
+    result[blockIdx.x + k] = s_tmp2[0];
+  }
+}
+
+// same as previous one but also scales the result by scaleFactors
+
+__global__ void MassIPV7part1withScaling(const double* __restrict__ u,
```
Comment (cnpetra) [inline src/LinAlg/hiop_cusolver_ir_custom_kernels.cu:95]: is this function used? if not, please comment out

File: src/LinAlg/hiopLinSolverSparseCUSOLVER.hpp [hunk 3]
```diff
@@ -70,242 +70,282 @@
...
+
+  int initializeKLU();
+  int initializeCusolverGLU();
+  int initializeCusolverRf();
+
+  int refactorizationSetupCusolverGLU();
+  int refactorizationSetupCusolverRf();
+
+  void IRsetup();
+  friend class hiopLinSolverNonSymSparseCUSOLVER;
```
Comment (cnpetra) [inline src/LinAlg/hiopLinSolverSparseCUSOLVER.hpp:209]: remove/comment it out? (three occurrences)

File: src/LinAlg/hiopLinSolverSparseCUSOLVER.hpp [hunk 4]
```diff
@@ -70,242 +70,282 @@
...
+  // matrix data of L and U factors
+  double* d_Lx_;
+  int* d_Lp_;
+  int* d_Li_;
+
+  double* d_Ux_;
+  int* d_Up_;
+  int* d_Ui_;
+
+  // matrix CPU data - this is needed to avoid allocing over and over and over
```
Comment (cnpetra) [inline src/LinAlg/hiopLinSolverSparseCUSOLVER.hpp:324]: typo? "allocing"

**Review by cnpetra — CHANGES_REQUESTED** [review]
PR looks good. Just left a couple of small comments. 

It would benefits to factor out the fgmres code as a `hiopKrylovSolver` at some point.

File: src/LinAlg/hiop_cusolver_ir_custom_kernels.cu [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  if(i < N) {
+    double temp = 0.0f;
+    for(int j = 0; j < k; ++j) {
+      temp += x_data[j * N + i] * s_alpha[j];
+    }
+    y_data[i] -= temp;
+  }
+}
+
+void massInnerProductTwoVectors(int n,
```
Comment (cnpetra) [inline src/LinAlg/hiop_cusolver_ir_custom_kernels.cu:195]: - change names of the functions accordingly to the coding styleguide, for example `mass_inner_product_two_vectors`

- for the two kernels, I would use  the "_kernel" suffix (but up to you)

File: src/Utils/hiopOptions.cpp [hunk 6]
```diff
@@ -902,6 +902,48 @@ void hiopOptionsNLP::register_options()
...
   }
+  // cusolver_lu inner iterative refinement options
+
+  {
+    vector<std::string> range = {"no", "yes"};
+    auto default_value = range[0];
+    register_str_option("ir_inner_cusolver",
+                        default_value,
+                        range,
+                        "Follow cusolver_lu  with FGMRES iterive refinement. "
```
[the remark this answers is not in the document]
Comment (kswirydo) [reply src/Utils/hiopOptions.cpp]: changed

File: src/Utils/hiopOptions.cpp [hunk 7]
```diff
@@ -902,6 +902,48 @@ void hiopOptionsNLP::register_options()
...
+                        1000,
+                        "FGMRES maximum number of iterations (default is 50). ");
+{
+    vector<std::string> range = {"mgs", "cgs2", "mgs_two_synch", "mgs_pm"};
+    auto default_value = range[0];
+    register_str_option("ir_inner_cusolver_gs_scheme",
+                        default_value,
+                        range,
+                        "Gram-Schmidt orthogonalization version for FMGRES. "
+                        "mgs: modified Gram-Schmidt (textbooki, default). "
```
[the remark this answers is not in the document]
Comment (kswirydo) [reply src/Utils/hiopOptions.cpp]: changed, should be "textbook" obviously

File: src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp [hunk 8]
```diff
@@ -875,737 +959,555 @@ namespace hiop
...
+        i++;
+        it++;
+        // Z_i = (LU)^{-1}*V_i
+        cudaMemcpy(&d_Z_[i * n_], &d_V_[i * n_], sizeof(double) * n_, cudaMemcpyDeviceToDevice);
+        cusolverRfSolve(cusolverrf_handle_, d_P_, d_Q_, 1, d_T_, n_, &d_Z_[i * n_], n_);
+        cudaDeviceSynchronize();
+        // V_{i+1}=A*Z_i
+        cudaMatvec(&d_Z_[i * n_], &d_V_[(i + 1) * n_], "matvec");
+        // orthogonalize V[i+1], form a column of h_L
+        GramSchmidt(i);
```
[the remark this answers is not in the document]
Comment (kswirydo) [reply src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp:1143]: yes and no;  the last one (GMRES-PM) is an unpublished result. I can add descriptive comments in the code, is this enough?

[the remark this answers is not in the document]
Comment (cnpetra) [reply src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp:1143]: that would be great, thanks

File: src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp [hunk 9]
```diff
@@ -119,8 +119,78 @@ namespace hiop
...
+    use_ir_ = "no";
+    use_ir_ = nlp_->options->GetString("ir_inner_cusolver");
+    if(use_ir_ != "yes" && use_ir_ != "no") {
+      nlp_->log->printf(hovWarning, 
+                        "Inner iterative refinement: %s is wrong. Use yes/no. Switching to default (no) ...\n",
+                        use_ir_.c_str());
+      use_ir_ = "no";
+    }
+
+    if(use_ir_ == "yes") {
```
Comment (nychiang) [inline src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp:132]: see my comment in 'hiopOptions.cpp'

File: src/LinAlg/KrylovSolverKernels.cu [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+// services by trade name, trademark, manufacturer or otherwise does not
+// necessarily constitute or imply its endorsement, recommendation, or favoring
+// by the United States Government or Lawrence Livermore National Security,
+// LLC. The views and opinions of authors expressed herein do not necessarily
+// state or reflect those of the United States Government or Lawrence Livermore
+// National Security, LLC, and shall not be used for advertising or product
+// endorsement purposes.
+
+/**
+ * @file KrylovSolverKernels.cu
```
Comment (nychiang) [inline src/LinAlg/KrylovSolverKernels.cu:49]: Should we add this file, `KrylovSolverKernels.cu`, into [CMakeLists.txt](https://github.com/LLNL/hiop/blob/e3e74ca6eeba97d46e6fdab56eb1abb55f7a3e54/src/LinAlg/CMakeLists.txt)?

**Review by nychiang — COMMENTED** [review]
LGTM. Thanks a lot! @kswirydo @pelesh

**Comment by pelesh** [comment]
> @kswirydo BTW, the latest commit fails on lassen CI. The error message is
> 

It was a missing brace. I think I fixed it.

**Comment by nychiang** [comment]
Some other errors: 
```
[ 32%] Building CUDA object src/LinAlg/CMakeFiles/hiopLinAlg.dir/hiopVectorIntRaja.cpp.o
/usr/workspace/hiop/gitlab_ci/lassen/189649/src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp(1084): error: identifier "use_ir_" is undefined
/usr/workspace/hiop/gitlab_ci/lassen/189649/src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp(1085): error: identifier "nlp_" is undefined
/usr/workspace/hiop/gitlab_ci/lassen/189649/src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp(1093): error: identifier "refact_" is undefined
/usr/workspace/hiop/gitlab_ci/lassen/189649/src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp(1094): error: identifier "ir_" is undefined
/usr/workspace/hiop/gitlab_ci/lassen/189649/src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp(1338): error: identifier "nlp_" is undefined
[ 32%] Built target hiopOptimization
```
I noticed that `use_ir_` is defined in `hiopLinSolverSymSparseCUSOLVER` but not in `hiopLinSolverSymSparseCUSOLVERInnerIR`.

File: src/Utils/hiopOptions.cpp [hunk 11]
```diff
@@ -902,6 +902,48 @@ void hiopOptionsNLP::register_options()
...
+                        1e-12,
+                        1e-16,
+                        1e-1,
+                        "FGMRES tolerance (default is 1e-12). ");
+
+    register_int_option("ir_inner_cusolver_maxit",
+                        50,
+                        1,
+                        1000,
+                        "FGMRES maximum number of iterations (default is 50). ");
```
[the remark this answers is not in the document]
Comment (kswirydo) [reply src/Utils/hiopOptions.cpp:934]: In my code, when it is "on", the memory (Krylov space etc) gets allocated. So before you check the iterations, you already wasted time and resources. I would need to modify the logic a little bit, can do this if needed
.

[the remark this answers is not in the document]
Comment (kswirydo) [reply src/Utils/hiopOptions.cpp:934]: okay, this was easy, changed it.

File: src/LinAlg/KrylovSolverKernels.cu (hunk 10 above)
Comment (kswirydo) [reply src/LinAlg/KrylovSolverKernels.cu:49]: I did.

**Comment by kswirydo** [comment]
I (1) added all the changes requested, (2) removed the "ir_inner_cusolver" from all relevant files,(3) added the KrylovCudaKernels.cu to CMakeLists. It compiles and links without any issues on PNNL marianas (with cuda 11.4). However, I am getting seg fault with `Address not mapped (1)` while running nlpSparseEx1, there is no hiop output and no readable trace. Will continue working on it. @pelesh @nychiang

**Comment by kswirydo** [comment]
Over and done. All changes have been applied, and the code compiles and runs correctly now. The branch is ready to merge. @pelesh @nychiang @cnpetra

**Review by cnpetra — APPROVED** [review]
looks great!

**Review by nychiang — APPROVED** [review]

**Pull request merged**